File: app/src/main/java/workshop/akbolatss/tagsnews/screen/reminders/RemindersAdapter.java
package workshop.akbolatss.tagsnews.screen.reminders;


import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import workshop.akbolatss.tagsnews.R;
import workshop.akbolatss.tagsnews.repositories.source.ReminderItem;

public class RemindersAdapter extends RecyclerView.Adapter<RemindersAdapter.RemindersVH> {

    private List<ReminderItem> mList;

    private final OnReminderClickInterface mClickInterface;

    public RemindersAdapter(OnReminderClickInterface mClickInterface) {
        mList = new ArrayList<>();
        this.mClickInterface = mClickInterface;
    }

    public interface OnReminderClickInterface {

        public void onReminderOptions(ReminderItem rItem, View view, int position);
    }

    private final View.OnClickListener mInternalListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            ReminderItem rItem = (ReminderItem) view.getTag();
            mClickInterface.onReminderOptions(rItem, view, 2);
        }
    };

    @Override
    public RemindersVH onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater mLayoutInflater = LayoutInflater.from(parent.getContext());
        View view = mLayoutInflater.inflate(R.layout.rv_reminder_item, parent, false);
        return new RemindersVH(view);
    }

    @Override
    public void onBindViewHolder(RemindersVH holder, int position) {
        ReminderItem rItem = mList.get(position);
        holder.bind(rItem);

        holder.mImgButton.setOnClickListener(mInternalListener);
        holder.mImgButton.setTag(rItem);

        holder.mCb.setOnClickListener(mInternalListener);
        holder.mCb.setTag(rItem);
    }

    public void onAddItems(List<ReminderItem> rItems) {
        if (rItems != null) {
            mList.clear();
            mList.addAll(rItems);
            notifyDataSetChanged();
        }
    }

    public void onAddItem(ReminderItem rItem) {
        if (rItem != null) {
            mList.add(rItem);
            notifyDataSetChanged();
        }
    }

    @Override
    public int getItemCount() {
        if (mList == null) {
            return 0;
        }
        return mList.size();
    }

    public class RemindersVH extends RecyclerView.ViewHolder {

        @BindView(R.id.cbActivate)
        CheckBox mCb;
        @BindView(R.id.tvTime)
        TextView mTime;
        @BindView(R.id.imgOptions)
        ImageView mImgButton;

        public RemindersVH(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void bind(ReminderItem rItem) {
            mTime.setText(onTimeFixer(rItem.getHour()) + ":" + onTimeFixer(rItem.getMinute()));

            if (rItem.getIsActive()) {
                mCb.setChecked(true);
            } else {
                mCb.setChecked(false);
            }
        }

        private String onTimeFixer(int time) {
            if (time == 0) {
                return "00";
            }
            Pattern ptr = Pattern.compile("^(?:[1-9]|0[1-9])$");
            if (ptr.matcher(String.valueOf(time)).matches()) {
                return "0" + time;
            }
            return String.valueOf(time);
        }
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/reminders/RemindersAdapter.kt
package workshop.akbolatss.tagsnews.screen.reminders


import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.rv_reminder_item.view.*
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.model.dao.ReminderItem
import java.util.*
import java.util.regex.Pattern

class RemindersAdapter(private val mListener: ReminderListener) : RecyclerView.Adapter<RemindersAdapter.RemindersVH>() {

    private val mList: MutableList<ReminderItem> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RemindersVH {
        val mLayoutInflater = LayoutInflater.from(parent.context)
        val view = mLayoutInflater.inflate(R.layout.rv_reminder_item, parent, false)
        return RemindersVH(view)
    }

    override fun onBindViewHolder(holder: RemindersVH, position: Int) {
        val rItem = mList[position]
        holder.bind(rItem, mListener)
    }

    fun onAddItems(rItems: List<ReminderItem>?) {
        if (rItems != null) {
            mList.clear()
            mList.addAll(rItems)
            notifyDataSetChanged()
        }
    }

    fun onAddItem(rItem: ReminderItem?) {
        if (rItem != null) {
            mList.add(rItem)
            notifyItemInserted(mList.size - 1)
        }
    }

    override fun getItemCount(): Int {
        return mList.size
    }

    fun updateItem(rItem: ReminderItem, pos: Int) {
        if (rItem != null) {
            mList[pos] = rItem
            notifyItemChanged(pos)
        }
    }

    fun removeItem(rItem: ReminderItem, pos: Int) {
        if (rItem != null) {
            mList.removeAt(pos)
            notifyItemRemoved(pos)
        }
    }

    interface ReminderListener {
        fun onReminderSwitch(rItem: ReminderItem, isActive: Boolean, pos: Int)

        fun onReminderOptions(rItem: ReminderItem, view: View, pos: Int)
    }

    inner class RemindersVH(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(rItem: ReminderItem, mListener: ReminderListener) {
            itemView.tvTime.text = onTimeFixer(rItem.hour) + ":" + onTimeFixer(rItem.minute)
            iconState(rItem.isActive)

            itemView.imgOnOff.setOnClickListener {
                mListener.onReminderSwitch(rItem, !rItem.isActive, adapterPosition)
                iconState(rItem.isActive)
            }
            itemView.imgOptions.setOnClickListener {
                mListener.onReminderOptions(rItem, it, adapterPosition)
            }
        }

        private fun iconState(isActive: Boolean) {
            if (isActive) {
                itemView.imgOnOff.setImageResource(R.drawable.ic_timer_24dp)
            } else {
                itemView.imgOnOff.setImageResource(R.drawable.ic_timer_off_24dp)
            }
        }

        private fun onTimeFixer(time: Int): String {
            if (time == 0) {
                return "00"
            }
            val ptr = Pattern.compile("^(?:[1-9]|0[1-9])$")
            return if (ptr.matcher(time.toString()).matches()) {
                "0$time"
            } else time.toString()
        }
    }
}
